tests/utils/tensorflow/test_rasa_layers.py

```diff
@@ -0,0 +1,855 @@
+            ],
+        ),
+        # Sequence layer with mixed features, hidden layers, no transformer, no MLM
+        (
+            RasaSequenceLayer,
+            model_config_basic,
+            {"attribute_signature": attribute_signature_basic},
+            attribute_features_basic,
+            [
+                [batch_size, max_seq_length + 1, units_hidden_layer],
+                [batch_size, max_seq_length + 1, units_hidden_layer],
+                [batch_size, max_seq_length + 1, 1],
+                [0,],
+                [0,],
+                [0,],
+            ],
+            "same_as_train",
+        ),
+        # Sequence layer with mixed features, no hidden layers, no transformer, no MLM
+        (
+            RasaSequenceLayer,
+            model_config_basic_no_hidden_layers,
+            {"attribute_signature": attribute_signature_basic},
+            attribute_features_basic,
+            [
+                [batch_size, max_seq_length + 1, units_concat],
+                [batch_size, max_seq_length + 1, units_concat],
+                [batch_size, max_seq_length + 1, 1],
+                [0,],
+                [0,],
+                [0,],
+            ],
+            "same_as_train",
+        ),
+        # Sequence layer with only sequence-level sparse features & MLM (to check the
+        # shape of token_ids)
+        (
+            RasaSequenceLayer,
+            model_config_transformer_mlm,
+            {
+                "attribute_signature": {
+                    SEQUENCE: [feature_signature_sparse_1],
+                    SENTENCE: [],
+                }
+            },
+            ([feature_sparse_seq_1], [], sequence_lengths,),
+            [
+                [batch_size, max_seq_length, units_transformer],
+                [batch_size, max_seq_length, units_hidden_layer],
+                [batch_size, max_seq_length, 1],
+                [batch_size, max_seq_length, 2],
+                [batch_size, max_seq_length, 1],
+                [
+                    num_transformer_layers,
+                    batch_size,
+                    num_transformer_heads,
+                    max_seq_length,
+                    max_seq_length,
+                ],
+            ],
+            [
+                [batch_size, max_seq_length, units_transformer],
+                [batch_size, max_seq_length, units_hidden_layer],
+                [batch_size, max_seq_length, 1],
+                [0,],
+                [0,],
+                [
+                    num_transformer_layers,
+                    batch_size,
+                    num_transformer_heads,
+                    max_seq_length,
+                    max_seq_length,
+                ],
+            ],
+        ),
+    ],
+)
+def test_correct_output_shape(
+    layer_class: Callable[..., Any],
+    model_config: Dict[Text, Any],
+    layer_args: Dict[Text, Any],
+    layer_inputs: List[  # TODO really use List? Should be Tuple of variable length...
```

Comment:
I test three different layers at once, each consuming the inputs as a tuple of a specific length (e.g. `CombineSparseDenseFeatures` takes a tuple with 2 elements). Can I just ignore these specific lengths and use `List` in the annotation (to allow various lengths)? The other option is to enumerate the different tuple lengths like:
```python
Union[
  Tuple[List[Union[tf.Tensor, tf.SparseTensor]]],
  Tuple[List[Union[tf.Tensor, tf.SparseTensor]], List[Union[tf.Tensor, tf.SparseTensor]], tf.Tensor]
]
```

Reply:
I think it's fine to use List here, the Tuple annotation is pretty verbose

Reply:
Okay, thanks 🙂